### ISSUE TITLE
Proposal: Interpolate statistics of anonymous visitors

### ISSUE THREAD

**Issue by tillsanders** [offen/offen#584]
This is a proposal to eliminate the blind spot in the analytics without compromising the offen concept.

I'm not familiar with the code base, so please understand this as a high level proposal without implementation details.

## Definitions

I really like the offen approach to privacy. I like the transparency and control it gives to visitors. At the same time, I see solutions like Plausible and Fathom become increasingly popular, because they track every visitor in a way that is at least privacy friendly. I like to think of them as *Anonymous Trackers*. Offen is what I, for the purpose of this proposal, would describe as a *Fair Tracker*. Now *Fair Trackers* are opt-in which is gives total control to the users. If users don't give consent, they will not be tracked at all. *Anonymous Trackers* will track every user and even employ methods such as subdomain forwarding to prevent anti-tracking methods. The do, however, go to great lengths to anonymize users. They keep little data and use hashing to protect personal information.

## Background

As a web developer and a member of the open source community, I value the privacy of my users. As a business owner (without unlimited resources for user research, focus groups, A/B testing, etc.) I need actionable usage statistics to understand how users interact with my product in order to improve UX. Users have a right to control their own data, but as a creator, I too have a legitimate interest to learn how many users I have. So I'm looking at solutions like Offen, Plausible and Fathom and try to wage the benefits. *Anonymous Trackers* can only provide limited insight because they need to be technically unable to connect too many dots. Their concept is to track all users, but not individually. *Fair Trackers* can potentially provide deeper insight, but can only track a possibly quite small subsample of users, because they are opt-in only.

Products like Google Analytics are on the extreme end of the privacy spectrum. They care only about the business owners. Offen however, ist also an extreme solution, just on the other end. Google gave full control only to business owners, Offen only to users. Solutions like Plausible and Fathom are somewhere in between, albeit obviously more on the privacy-friendly side. I really like what Offen is building, but I'm afraid the approach might be too restrictive to find wide commercial usage. With this proposal, I seek to provide just enough data to satisfy the cravings of business owners without compromising the Offen concept and values.

## Proposal

I would love to see Offen implement a feature that would allow interpolation of actual usage, based on a) the data collected using the opt-in tracking and b) additional, but very limited data about all visitors. 

The additional tracking proposed would be obligatory for all visitors, but would be cookie-less and would only collect:

- a timestamp of the moment a request was made
- the requested domain + path

This is very little and it is not personal data. But this information could be combined with the rich information about consenting users to interpolate actual usage. The way Offen currently collects statistics, I know how many users opt-in to analytics, but I don't know how many users don't. Collecting just this additional data would enable the following example.

### Example

On a given day, I have 1000 users on my page. 70 of which visit a certain subpage. Only 10% of all users opt-in to analytics. So in my auditorium, I will only see 100 users in total on that day and 7 on that subpage. That gives me a false sense of usage. My site is actually 10 times more popular than I think. If I were to sell ads on my page, to maybe make my project sustainable – even if I want to do it in the good old-fashioned, zero-tracking way, Offen would make it hard to determine a fair price because I don't know the total number of views.

However, if Offen were to collect the additional tracking data of every user as described in the proposal, the 10% of users that share their statistics with me, could be interpolated to show me an approximation of the real usage. That way Offen could tell me that there were at least 7 visits on that subpage, but based on the additional tracking, there probably were actually 60 - 80 visits.

This could be visualised using a box plot or similar.

## Rough ideas for implementation

1. Many websites collect logfiles for various reasons. When it comes to DDOS protection, logfiles are a valuable tool. Offen could read the logfiles and gather only the information needed. The logfiles could (as they should be anyway) also be anonymised and short-lived. Offen could also further aggregate the data collected.
2. Offen could collect the needed data itself, similar to the way Plausible and Fathom are doing it.

Personally, I would prefer the first idea. Also, the feature should be optional. 

## Advantages

- more accurate statistics that are far closer to the real numbers
- more appeal to businesses
- makes it easier to sell ads without resorting to ad networks which probably always compromise user privacy
- many websites collect logfiles anyway, even if just anonymised, ome might even use tools like GoAccess to complement the Offen stats
- only very little additional data would be collected and no personal data

## Disadvantages

- the concept is more difficult to understand for non-technical folks – if not done well, it might appear like a cheap trick
- with either solution, the setup would be a little more complicated (another reason, this feature should be optional)
- analytics are often hosted separately from where the logfiles live, so to grant access, logfiles might be exposed by accident

## Conclusion

I believe this feature would be a great addition and would improve Offen's competitiveness towards the *Anonymous trackers*. I would even argue, that this approach would be fair not only towards users, but also towards projects and businesses.

**Comment by m90** [comment]
Thanks for this very detailed proposal and your thoughts around anonymous trackers, fair trackers and who needs what. It's interesting to see such things spelled out and it's also very good to be reminded of the business perspective that many of our potential users have.

This will be a lengthy answer with a lot of anecdotes and no real conclusion (I'm delivering the letdown upfront), but here we go:

## Offen in Alpha - "Plus" feature

When we started working on Offen something around 2 years ago, the very first version included a concept that was eerily similar to what you are proposing: We called it "Plus". It worked in the following way:

- a user that had opted out would behave like today
- a user that had opted in would behave like today
- a user that had no consent decision yet or who would have dismissed the consent banner would generate something we called a __anonymous event__ on each pageview. This anonymous event would contain nothing but a timestamp (your suggestion also adds domain and path).

In the UI we would then display a value called "Plus" which would basically do what you have described: display the ratio of anonymous events in your current query. In the long run we wanted to also add a toggle that would let you multiply your actual stats with this factor, displaying interpolated values, but since we abandoned this behavior at some point we never made it to this interpolated dashboard.

From a technical perspective, this feature would still be kind of easy to add to the existing codebase as of today. I would assume some hooks for this behavior are still left dangling in there, gone unnoticed in some cleanup (the last PR removing such places dates back to January 2021: https://github.com/offen/offen/pull/542).

In case you're interested, you can clone the repository and checkout a state from sometime around August 2019 (1a378db340742bedba4381e59a9ea0e03a186e3e maybe?) and poke around in the past.

## Focusing on self-hosting

At that point in time, Offen was still going to be an Open Source SaaS kind of thing we'd run ourselves. Looking for some funding we ended up getting in touch with NLNet, who as you might know, still fund development of Offen today (thank you!). This had two effects on the direction of the project: we agreed to making Offen software that can be self hosted, and (this is the important part for this issue) we agreed that the first version of Offen would be Opt-In only, no exceptions.

At that point we were still thinking to run the SaaS version including the "Plus" feature in parallel to developing the _Fair_ self hosted version, but while doing this we noticed some other influences on the project taking over.

## Focusing on being Fair

So we went ahaead and developed the _Fair_ version of Offen and the following things happened:

### Do you really need the exact number of visitors?

Of course we dog food Offen ourselves and we found that in basically all scenarios we don't really need to know the exact number of visitors. We're interested in seeing the usage patterns, and we see these using Offen too. We also ran a comparison with GoAccess log analytics for a while and while the numbers were indeed different, the patterns weren't. This laid back attitude might of course not apply to projects that are more business driven where hard numbers matter. Then again, who might know which number is a bot and how exact those `hash(ip, ua, daily_salt)` solutions used by Fathom et al really are when it comes to distinguishing users?

### It's interesting to explore this extreme end of the privacy spectrum

Offen is not a business and will never become a business. Receiving the funding we get from NLNet allows us to be a (annoying at times) thought experiment as much as we are Free Software. It definitely hinders adoption (this would be insane when done by a business, but it turned out we aren't), but raising the question "Let's think through what doing web analytics in a really fair way means, and then let's reevaluate if you really need web analytics." is very important to us. If we make someone say "You know what, I'm not gonna apply analytics at all because of privacy implications", this means just as much as a proper install for us as a project.

## Bringing our ideas to the real world

Your proposal comes at an interesting time as at this very moment, our work starts diversifying and we do not focus on Offen as a software exclusively anymore, but we also think about ways to make our ideas and principles available to other projects. We are working on standardized ways to describe analytics interfaces and allowing others to apply the data exchange protocols we use (N.B.: if you're interested in this, we'd be happy to share the work in progress with your for feedback).

And just like you propose, it would be very interesting however to bring Offen itself closer to the real world too. Right now, I'm just not entirely sure how this could work.

## Requirements for implementing such a feature (again)

To wrap things up, let me add some more thoughts about what requirements there would be for adding back the "Plus" feature or something similar in the Offen of today.

### Anonymous data stays anonymous

We would like to avoid Fathom/Plausible/YouNameIt style server side hashes for multiple reasons. Two unwritten laws in the Offen codebase are: "you never look at Request IPs" and "you never look at UA Strings". This might seem stupid when doing things like rate limiting or similar, but we can still get things working as intended, and it helps us remain in perfect posture when it comes to handling potentially sensitive user data.

In addition to that I would argue the server side hash approach is not entirely anonymous. The space of IP / UA-String combinations is finite, which would allow anyone who gets hold of the salt in use to perform reidentification attacks as described here: https://edps.europa.eu/data-protection/our-work/publications/papers/introduction-hash-function-personal-data_en - it's probably not worth the computing power wasted compared to the loot an attacker would make, but then this is probably also more about keeping that perfect privacy posture upright.

We could however, indeed send an anonymous event as already described in the "Plus" feature from the client

Logs like you propose sound good to but raise a few problems I'm not entirely sure how to solve yet. They would not allow us to differentiate consent state easily (I think Opt-Out needs to mean Opt-Out of everything for us). Also they are terribly tricky when it comes to prevent accidentally leaking information in them, see https://github.com/offen/offen/issues/455 and https://github.com/offen/offen/pull/472 for example. Then again, I have not put too much thoughts on how logs could help us handle this yet, but I will definitely spend time thinking about it.

### Avoid option proliferation

It might be tempting to say "we could be adding a `OFFEN_APP_COLLECT_ANONYMOUS=true` setting" and keep the current default behavior, but I don't think this is what we want. Unlike other analytics software we are also user facing, and this means we need to be able to offer a consistent story for users. If users learn that there are multiple ways the software that asks you to opt-in before doing anything works, they will lose trust and interest immediately and we'd have the ship sinking in no time.

This means, if we'd implement such a feature, it would have to be the only way Offen can behave, and it would also need to be understandable for users.

### The limbo of pending consent

As a preface - it's probably obvious - but I will just repeat it: an Offen without Opt-In and Opt-Out is not going to happen. It is a core part of our value proposal and we also think it makes sense. If consent is not someone's cup of tea, there are already lots of solutions out there.

This is probably the trickiest part to solve here. We do know already that users who have opted in are ok with their usage data being collected and that users who have opted out do not want any data recorded (they would also need to be exempt from anonymous data if we stick to our value proposal). So what is left for us to define is how to handle users that do not make a consent decision or dismiss the banner.

We could collect anonymous events for such users without having to set a cookie or similar (i.e. it would be ok with regulations), but there are some questions about this that I don't know how to handle yet:

- assuming we send an anonymous event before a user opts out, how do we delete this event again? This could possibly be worked around using navigator.sendBeacon and deferring the transmission, but I don't know the full implications of this right now.
- the same as above, but the user opts in. What happens to the anonymous event?
- much more complicated than the technical parts: how do we explain this to users? You could of course argue: well it's not much different than triggering a log line which would happen anyways, but we'd need to explain that there are two different kinds of "No" here which is probably very hard.

Collecting anonymous events for users that have opted out as well seems like a bad decision to me. No means no.

---

So this was a lengthy explanation without any conclusion other than "It's complicated", yay! You might have noticed however that your proposal scratched an itch and the direction to which you are pointing sounds interesting for us as well. We'll discuss this further internally again, and we would also like to invite you to feedback to anything I just have written down here, these kinds of conversations are extremely valuable for us. Thank you!

**Comment by m90** [comment]
I think I was missing an important detail in my comment about that "Plus" feature above: back then users still could dismiss the consent banner without making a consent decision. We changed that in order to get it "out the way" so that site owners can move on with their stuff afterwards.

**Comment by tillsanders** [comment]
Thanks for the detailed reply, Frederik!

You provided a lot of insight and I would like to elaborate on a few:

## Differences between the original 'Plus' feature and my proposal

Very interesting read! I can already imagine what this would have looked like. I like the concept – too bad, it was never completely implemented. While it is quite similar to what I had in mind, there is also a big difference in the way users are counted (I'm not using the word 'tracked' here on purpose, but more on that later). The Plus feature never counted users that had opted out. My proposal would count every user because I think that's the only way to deliver the benefit that I'm looking for with this proposal. Say, 10 % of my visitors opt in, 60 % opt out and 30 % don't provide any decision: with the original Plus feature there is still a discrepancy of 30 % which I would like to eliminate. But I think that is the only important difference. Other than that, my proposal would also use domain and path, but I think that would be totally debatable. 

## Thoughts on SaaS and self hosting

I think allowing people to self-host Offen is an important step and I appreciate this! In fact, it is something I'm missing from Plausible and Fathom. Obviously, self-hosting an analytics server can become difficult if you have a lot of traffic (see the recent [blog post on scaling Fathom](https://usefathom.com/blog/worlds-fastest-analytics)) so as a project it is extremely difficult to provide both SaaS and a self-hosted solution because at some point the surrounding services (especially the database) will become quite different. A small instance will be perfectly fine with a simple MySQL database, whereas the SaaS instance quickly outgrows that. Not that it's not possible to provide both, but for a small team it makes perfect sense to focus on one thing. Also, for commercial growth, SaaS is the obvious choice. So to wrap this up, I think going self-hosted is the right way for Offen at this time. 

On the other hand: We also have a small development team and while we would love to have our own GitLab instance, container registry, npm registry, mail server and analytics server, maintaining all this is just not cost effective. The hours we would need to invest to host all these services (properly, with backups and stuff) are just too expensive compared to SaaS solutions that are essentially worry-free. 

So this is quite off-topic, but to wrap things up: I think going self-hosted was a good decision! But especially small projects and businesses can profit from SaaS. So maybe when Offen grows, there will be room for a hosted service. Maybe this could be implemented not as a normal SaaS, but more like Laravel Forge – a service that will help you manage our own Offen instance. That way, people would still be in control of their data while not having to do a lot of maintaining and you wouldn't need to think about scaling.

## Do you really need the exact number of visitors? – No, but also yes.

You are obviously right about the patterns. I don't need to track all users to be able to make deductions about my product. 

1. But yes, for **businesses** hard numbers are relevant. For one, many managers are simply too fascinated with them, but also investors can use numbers like these as key metrics.
2. Again, **advertisements**. I'm not a fan of ads personally, because they are just a waste of traffic and cripple otherwise beautiful websites, all while also tracking me. Further, I'm not a fan of ads professionally, because I feel like their value in growing a business is overrated these days. But! Especially medium sized content creators can use them for monetization. But today this either involves signing a contract with some silicon valley devil or a more privacy focused ad network. Both require some form of payment. So if I were in such a position, I would be looking for long-term promotions. But to sell these to interested companies, I obviously need accurate numbers. So that's a probable use case.
3. **Small projects:** I have a personal website and a small open source project. Both with very little traffic. Just a few visits per day. I'm expecting this to stay that way and I don't care too much. But I do find it motivating to see people visiting the website of my open source project. Every now and then I take a look at [the stats](https://plausible.io/open-election-compass.com), just to remind me that there are people interested in this. Just a few. On that site people would probably be more willing to opt in, but I assume I would still miss most of them.

So, you're generally right. I don't need the exact count to improve my product, but there are cases where the exact number is still very interesting.

## On hashes as a means of identifying users

That might have been a misunderstanding. I think the idea as used by Fathom and Plausible is solid, but I wasn't suggesting implementing a similar technique in Offen.

## The extreme end of the privacy spectrum

So this is the part where you basically explain that my proposal is simply out of scope. And that's obviously okay. It's important to have a Not-To-Do-List. So counting the exact number of users is not a feature Offen provides. Taking privacy, consent and transparency to the extreme is the idea behind Offen, so if you feel like the proposed feature is not promoting the goal or is even in contradiction, then you shouldn't do it. 

I especially understand the "problem" of the coherent user story. This is completely reasonable. I now agree, a feature like the one proposed should not be optional.

### There is a difference between tracking and counting users

While reading your reply, I noticed that there is a difference between tracking and counting users. 

I want to know how people interact with my product. I want to know where they come from, what content and features are the most interesting, how often they return, what browser they use, etc. But I don't want to track anyone without their consent, because in my understanding, the power of web analytics is unique. When I build a physical product, I cannot gather this data automatically. If I want to know how people interact with a physical product, I have to ask them. And if they don't want to tell me, then I have to respect that. So just because web analytics makes it possible to track usage like never before, I can't simply take that information, because I wouldn't be able to do that with a physical product either. 

But! Even with physical products, I know how many I sell. When I'm serving a customer in a restaurant, I usually don't know anything about them, but I do know how many plates I'm washing. When I'm selling tickets for a play, I don't know who I'm selling them to, but I do know how many seats I have left.

That's the subtle difference between tracking and counting. I don't want to propose tracking people without consent. I just want to count them. Counting people is probably the tiniest bit of information I can have about a person and that's really all I would like to add to Offen with this proposal.

Let me add a rather philosophical thought to this: The mere existence of not even a specific person, but just ... someone is not personal data. It's not tracking, the way I see it. A simile: if I were to track the load on my server's SSD, I could make assumptions about the visitor count at a given point in time. If I wanted, I could feed this information to Offen. Would you consider this tracking? The result would (apart from a loss in accuracy) not be so different. Yet all I do is reading the load of a component in my server.

### Might the extreme approach be counter-productive?

My business currently uses Fathom on various sites. Though Fathom and Plausible don't stress that too much, their numbers can be expected to be more accurate than Google Analytics or Matomo (seen that myself). And I'll be honest – that and the zero-cookie solution were of great appeal to me. I want to respect the privacy of my users, but I also need accurate data and they deliver not only that, but also good arguments that calm my consciousness. But I'm still following Offen to see where this is going to maybe find a better alternative for everyone. But while I consider moving my personal projects to Offen, despite the disadvantage concerning accurate visitor count (see above), it is hard to do so for my business. Because tools like Facebook Analytics, Google Analytics and E-Mail Tracking exist, at least for now. It's not my decision alone and it is hard to argue against their use when it is industry standard and there is no alternative solution that meets all our needs. This is the reality for many companies at the moment. On a side note, personally I'm grateful Apple is making these practices harder. Anyway, this leads to the following: I can use a more privacy-caring solution like Offen or Plausible and Fathom for that matter, but as a business that is app-based, I cannot ignore the benefits the industry standard offers. So I end up with a privacy friendly solution and the usual garbage on the same site. And that would also hurt the user story of Offen. Because Offen might respect the opt-out, but other tools might not. Or because Offen might allow data to be deleted, but obviously can only delete the data under its control.

Obviously, this is not a problem with Offen itself, but with the industry as a whole. But maybe, if Offen allowed counting all users, it would make it more competitive with industry standards and thus improve privacy for more users.

## Summary

I totally understand the problems you have with this proposal and I agree on many, if not all points you made. I still think counting, not tracking all users could be reconsidered. I really like my analogy to physical products, as it both strengthens your stance on consent, as well as my argument in favour of counting all visitors. But if you decide this would not meet the values and goals you have set for Offen, then you should put those first. Open source strives the most when combined with strong beliefs and visions! Thank you for your valuable insight and transparency. This is really interesting!

---

## Moving forward: On standardisation

### Privacy Labels

While explaining my problem with multiple tracking tools on the same website with different standards regarding privacy, it came to my mind that a standardised labelling system would be interesting. A bit like what Apple does now. Websites could use these to communicate what they do with your data. This would be far quicker to understand than lengthy privacy policies. But it would, of course also need to find great adoption and standardisation and I'm almost certain, that someone already attempted something like that and failed because the big players weren't involved. So ... still, would be cool.

### OpenElectionCompass

As I mentioned, I have a small open source project. It's a software to create voting advice applications like the Wahl-O-Mat or the Stemweijzer. I recently added a feature that allows users to donate their answers (basically tracking) to scientific research. Maybe I could use a standardised format for that such as the one you mentioned. Let's talk about that!

You can take a look at it here: <https://open-election-compass.github.io/client/demo-hybrid.html>
Opt-in at the end.

**Comment by m90** [comment]
Thanks for following up on my reply. I agree with basically all points you mention, and I think it all of this comes down to answering a single question.

## What does Opt-Out even mean?

> We only access usage data with your consent.

This is the wording Offen currently uses in the consent banner. The options given are "I allow" and "I don't allow". To me, this makes it kind of complicated to introduce the anonymous behavior we've been talking about for everyone: I just clicked "I don't allow", but I am still showing up in statistics?

> I don't want to propose tracking people without consent. I just want to count them.

It's true that you can always infer someone's presence from some signal and I would also think it's morally ok to just "count" your visitors without their consent. However, in the current state of Offen, we have created the situation where we actively ask for user's consent to do so. If they say no, we'll have to buy into that decision and simply never collect their usage data, even if they still leave traces of some sort. Differentiating _tracking_ and _counting_ gets too complicated when you make things end user facing.

It's insanely important to make the consent decision as clear as possible for us as otherwise we fall prey to the "oh this annoying cookie banner thing that wants to dark pattern me, go away" attitude that is out there. In earlier versions we had two sentences in the banner instead of one, and feedback on this version was already quite devastating (this is quoted as is from real persons):

> I guess they didn't even read the initial pop-up. Reading it again,
"transparent analytics" and "Help to make this website better" to me
signals "do not read any further" and "they are tracking me no matter
what I click".

So I think allowing users to opt out, but then still count them instead of tracking them doesn't really help build trust in any way, but will reinforce the "I'm being scammed no matter what" attitude out there. But we'd like Offen to show that collecting usage data does not necessarily mean scamming, it can be a fair and transparent process.

## Forking Offen and going crazy

So as we've seen Offen as is has a lot of constraints that are quite hard to overcome. Just as a thought experiment, let's assume someone forks Offen, rebrands it and and makes it more business friendly by adding the anonymous counting on top. How could this work?

### Keeping consent a yes/no decision

The new product would come without the baggage of a "no means no" consent mechanism, so one could make the "no" an "anonymous only". This would allow the product to both count and track and then come up with interpolated numbers. However, this will likely be interpreted as "they are tracking me no matter what I click".

### Making consent a yes-only decision

Another option would be saying consent is only about opting in to tracking. There is no way of saying "no" (you probably could dismiss the banner somehow still). This would still be "they are tracking me no matter what I click", but at least it'd be transparent about it.

### Removing consent entirely

Consent could also be removed entirely, while still allowing users to opt-out of tracking somehow while they would still be counted. This would create two problems:

- visibility: how do users even know their data is visible to them and they could opt out. Would there still be a banner, but without any actions to be taken?
- regulations: this is probably not GDPR compliant (complying with regulations has never been an objective of ours though, we want to build analytics software that caters the needs of users and operators alike).

## What's a software to do?

I feel thinking about such options in the scope of a fork is kind of interesting as it allows us to keep the story of Offen as it is intact, and writing this down also reassures me that it's a coherent one we should try to keep alive. If not for business reasons, then for idealistic ones.

I'm not entirely sure about the options I have just written down. They all have their pros, but they also have weak points that would still need to be resolved. Putting aside the feasability of "hey let's just run a fork, LOL" I will have to let this sink in and have my subconscious ponder the options for a while. If you have any immediate reactions here, please share them with us.

---

## Sideshows

As mentioned there are two side tracks we are working on / planning to work on:

### analytics.txt

This is supposed to be a specification for websites and services to disclose information about their usage of analytics software and other trackers, making it discoverable for both machines and humans. We have been working on this for the last few weeks and you can check out first versions of the draft here: https://github.com/offen/analyticstxt - we'd love to learn about your thoughts on this, but please keep in mind that this draft is still to be considered incomplete and is subject to rapid change before we submit it to the IETF (which will happen this summer).

### Offen protocol + OpenElectionCompass

Another project on the horizon is formalizing an "Offen protocol" that defines how clients and servers can share (usage) data in a secure and bidirectional way, formalizing what is happening in Offen at the moment. This will be something we tackle later in the year, but if it sounds interesting to you, please email us at <hioffen@posteo.de> and we can share further details with you.

**Comment by tillsanders** [comment]
Thanks for your continued replies ;)

> I understand the problems with my proposal and I agree. I normally don't change subjects in an issue, but I will make an exception as this has been more about the UX aspects and the philosophical dilemma anyways, which I find interesting to pursue.

## Summary

> I figured, since our posts are so long, it might be easier with a summary upfront ;)

I'm leaving my proposal behind but I'm still focusing on what Offen would need to provide to become a feasible alternative for medium sized projects and businesses. I want to argue that
- because of Offen's idealistic approach it cannot be used side by side with any other tool and 
- in order to achieve commercial or even more widespread use, Offen needs either more core features or a plugin API to make the use of additional tools unnecessary.

## Opt-out: No means no means no other tools beside Offen?

I see what you mean and I while I think there is a great difference between counting and tracking users, I agree that it is not possible to explain that to most people, especially in the incredible short time they grant "cookie banners".

Following these arguments, I'm coming to the conclusion that:

- On the **User side**: The "I'm being scammed" experience you mentioned can only be fought if Offen is the only tracking tool deployed on a site. If there are multiple tools like Offen + Facebook or even a privacy-friendly set like Offen + GoAccess, the user story will be compromised. So it really boils down to using only Offen and nothing else because adding even a single other tool makes things too difficult to understand for users.
- On the **developer side**: Offen promises to be 1) strictly opt-in, 2) completely transparent and allows 3) deletion of data. Integrating these central features with other tools like GoAccess, Facebook Analytics, Sentry or whatever is practically impossible. **So if I, as a developer, want to keep true on the promise I'm giving by using Offen, I can effectively not use any other tool on the same site.**

## Difficulties in commercial application

If I cannot combine Offen with any other tool for additional metrics because of the issues above, this means serious problems regarding commercial application. Even if my business or project is extremely progressive, doing without tools like Sentry is hard and not a good decision for a business. And what about referral tracking, which can be a valuable marketing channel for any project? There are legitimate use cases that become extremely difficult if I want to keep the "Offen promise".

## Solution: Decoupling the data layer / a plugin system / more core features?

Additional use cases like referrals, performance and error monitoring can only be handled without breaking the promise on privacy if Offen provided a plugin API or (for a consistent user story) more core features.

For example: I could build a plugin to handle client-side error monitoring. The plugin would only be executed in the browser by Offen, if the user provided consent. The additional data collected by the plugin would be saved in the Offen data store so it would be subject to the same rules as the "normal" tracking data and also displayed in the Auditorium.

Providing such functionality as a core feature, albeit an unusual approach, would probably be better for the user story. 

Would be interesting to know how many sites that use Offen also deploy other tools.

---

## On forking Offen

I don't think you should. Not even business-wise. You're making very thoughtful and good arguments for privacy and consent. Offen's approach, afaik, is unique because it doesn't compromise at all. Sure, implementing my original proposal would probably hit a nerve and wouldn't be too hard to position on the market. It would be like *Fathom Plus*™ or *More Plausible*™ because it would provide the same data and more thanks to some users who opt in. 

But with some more core features like the ones I touched on above, equally uncompromising, Offen would be more interesting. Maybe it's better to become less like the competition.

---

## analytics.txt

Nice project, that is basically what I had in mind with my suggestion to establish something like privacy labels. Now all it needs is an interface in the browser, maybe similar to what DuckDuckGo is doing.

## Offen protocol

The Offen protocol sounds a bit similar to what I suggested above but I'm not sure if that is the intention behind it or a possible way to use it? Is it just for requesting data that Offen collected or can it also be used to save additional data in Offen? If so, it might be interesting for my project as I have data that is specific to the OpenElectionCompass. Right now I'm working on some other features, but when I get back to the analytics stuff, I'll get back to you!

**Comment by m90** [comment]
> For example: I could build a plugin to handle client-side error monitoring. The plugin would only be executed in the browser by Offen, if the user provided consent. The additional data collected by the plugin would be saved in the Offen data store so it would be subject to the same rules as the "normal" tracking data and also displayed in the Auditorium.

This is a pretty interesting thought, and also something we talked about NLNet (in a slightly more convoluted way) already. We'll definitely keep such use cases in mind when we design the protocol we talked about already.

---

> Now all it needs is an interface in the browser, maybe similar to what DuckDuckGo is doing.

We were thinking building a browser extension - or even better: piggybacking on an existing browser extension - would be a good start here. Next step would be approaching search engine like tools that could consume and display such data to users. Is this what an "interface" would look like to you?

**Comment by tillsanders** [comment]
> This is a pretty interesting thought, and also something we talked about NLNet (in a slightly more convoluted way) already. We'll definitely keep such use cases in mind when we design the protocol we talked about already.

Nice, then I'll keep an eye out for that protocol ;)

> Is this what an "interface" would look like to you?

Pretty much, yeah! I like the way the DuckDuckGo browser plugin gives "grades" to sites and I think they would be able to make use of analytics.txt.

On another note: Lighthouse support would also be a great way to foster adoption.

**Comment by m90** [comment]
> Lighthouse support would also be a great way to foster adoption.

That sounds good, I haven't thought of it yet. I wonder if it'd be possible to do this by providing a plugin https://github.com/GoogleChrome/lighthouse/blob/master/docs/plugins.md
